### PR TITLE
Fixed #24997 -- Enabled bulk_create on a proxy model

### PR DESCRIPTION
**Pull request by wkschwartz**
This patch enables `SomeProxyModel.objects.bulk_create(objs)`, which previously raised a `ValueError`. The real difficulty with bulk insertion is the multi-table case, and this patch correctly detects when a proxy model inherits from some other multi-table models.

The tests pass on Mac OS 10.10.2, Python 3.4.3 with both Sqlite 2.6.0 and PostgreSQL 9.4.4.

See https://code.djangoproject.com/ticket/24997.

### CODE REVIEW

**Comment by wkschwartz** [comment]
Jenkins is upset about the sort order of the imports I added to `tests/bulk_create/tests.py`. I chose that order to keep classes together that are used together. If there is a strong preference for alphabetical ordering over topical ordering, I will make the change.

**Comment by charettes** [comment]
Hi @wkschwartz, all the Django modules import order follow the rule set we defined for isort. Please stick to it in the name of uniformity.

**Comment by wkschwartz** [comment]
@charettes,

No problem. Fixed in 3725f0591e.

I'm guessing you'll want these commits squashed before accepting the PR, but I didn't want to squash it until you had seen the diff of the last change. Let me know what else I can do to make the PR acceptable.

**Comment by charettes** [comment]
Squashing can wait final review. I'll run the test suite against your changes.

**Comment by charettes** [comment]
Buildbot, test this please.

File: django/db/models/query.py [hunk 1]
```diff
@@ -423,13 +423,18 @@ def bulk_create(self, objs, batch_size=None):
         # this by using RETURNING clause for the insert query. We're punting
         # on these for now because they are relatively rare cases.
         assert batch_size is None or batch_size > 0
-        if self.model._meta.parents:
-            raise ValueError("Can't bulk create an inherited model")
+        # Checking all the parents share the same concrete model with the our
+        # model detects the inheritance pattern ConcreteGrandParent ->
+        # MultiTableParent -> ProxyChild. Simply checking self.model._meta.proxy
```
Comment (timgraham) [inline django/db/models/query.py]: where there any test failures with the simpler solution? If not, it would be nice to add test for those cases. I saw one test in `tests/bulk_create/tests.py (test_inheritance)`

Comment (timgraham) [inline django/db/models/query.py]: nevermind, I see you did add them. I'd just suggest to use separate test methods for the valid and invalid cases.

File: tests/bulk_create/tests.py [hunk 2]
```diff
@@ -51,6 +54,30 @@ def test_inheritance(self):
...
+        self.assertQuerysetEqual(ProxyCountry.objects.all(), {
+            "Qwghlm", "Tortall"
+        }, attrgetter("name"), ordered=False)
+        ProxyProxyCountry.objects.bulk_create([
+            ProxyProxyCountry(name="Neitherlands", iso_two_letter="NT")
+        ])
+        self.assertQuerysetEqual(ProxyProxyCountry.objects.all(), {
+            "Qwghlm", "Tortall", "Neitherlands",
+        }, attrgetter("name"), ordered=False)
+        with self.assertRaises(ValueError):
```
Comment (timgraham) [inline tests/bulk_create/tests.py]: please use assertRaisesMessage to check this is the ValueError we expect.

File: tests/bulk_create/tests.py [hunk 3]
```diff
@@ -51,6 +54,30 @@ def test_inheritance(self):
             "Nicholas's",
         ], attrgetter("name"))
 
+    def test_proxy_inheritance(self):
+        """Ticket 24997: Support bulk_create on proxy models."""
```
Comment (timgraham) [inline tests/bulk_create/tests.py]: the ticket reference doesn't seem necessary since the expected behavior is straightforward

File: docs/releases/1.9.txt [hunk 4]
```diff
@@ -322,6 +322,9 @@ Management Commands
 Models
 ^^^^^^
 
+* :meth:`QuerySet.bulk_create() <django.db.models.query.QuerySet.bulk_create>`
+  now works on proxy models (:ticket:`24997`).
```
Comment (timgraham) [inline docs/releases/1.9.txt]: ticket reference not needed (only for minor release notes)

File: tests/bulk_create/tests.py (hunk 2 above)
Comment (wkschwartz) [inline tests/bulk_create/tests.py]: Other tests in this module use `assertRaises` instead of `assertRaisesMessage`. I agree that `assertRaisesMessage` (or `assertRaisesRegex`) is (are) better. Should I change every instance of `assertRaises` that tests this particular error, or only those instances of it added in this pull request?

Comment (timgraham) [inline tests/bulk_create/tests.py]: You may change the others if you like, but please make it a separate commit.

File: django/db/models/query.py (hunk 1 above)
Comment (wkschwartz) [inline django/db/models/query.py]: There's an invalid case in the middle of `test_inheritance` and two at the end of `test_proxy_inheritance`. The `assertQuerySet` should come after both the valid and invalid cases (see a687b0b61ea5c3). I'd rather change the name of `test_inheritance` to `test_multi_table_inheritance`, i.e., the tests are organized by multi-table versus proxy rather than by valid or invalid. Thoughts?

Comment (timgraham) [inline django/db/models/query.py]: I'd go with separate methods for each of the different combinations.

Comment (wkschwartz) [inline django/db/models/query.py]: See 2342115d56039fa3a6928db740334d2c56e859f5. Does that do the trick?

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/bulk_create/tests.py [hunk 5]
```diff
@@ -35,22 +38,48 @@ def test_efficiency(self):
         with self.assertNumQueries(1):
             Country.objects.bulk_create(self.data)
 
-    def test_inheritance(self):
+    def assert_inheritance_error(self, *args, **kwargs):
+        """
+        Like assertRaisesMessage but with the right arguments for an inheritance
+        error already curried.
+        """
+        return self.assertRaisesMessage(
```
Comment (timgraham) [inline tests/bulk_create/tests.py]: Now that I look at this, I think I would combine the 3 cases that use this in 1 test method to avoid the helper method. We prefer using the context manager version of `assertRaisesMessage` for better readability.

File: tests/bulk_create/models.py [hunk 6]
```diff
@@ -6,6 +6,28 @@ class Country(models.Model):
     iso_two_letter = models.CharField(max_length=2)
 
 
+class ProxyCountry(Country):
+
```
Comment (timgraham) [inline tests/bulk_create/models.py]: I think we can chop the blank lines before Meta in these classes

File: docs/ref/models/querysets.txt [hunk 7]
```diff
@@ -1768,6 +1768,12 @@ This has a number of caveats though:
   does not retrieve and set the primary key attribute, as ``save()`` does.
 * It does not work with many-to-many relationships.
 
+.. versionchanged:: 1.9
+
+    ``bulk_create`` now works with proxy models. Previously it worked for
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: Actually, I think "Support for using bulk_create() with proxy models was added." should suffice. I find the "Previously..." sentence a bit awkward and unnecessary.

File: django/db/models/query.py [hunk 8]
```diff
@@ -423,13 +423,18 @@ def bulk_create(self, objs, batch_size=None):
         # this by using RETURNING clause for the insert query. We're punting
         # on these for now because they are relatively rare cases.
         assert batch_size is None or batch_size > 0
-        if self.model._meta.parents:
-            raise ValueError("Can't bulk create an inherited model")
+        # Checking all the parents share the same concrete model with the our
```
Comment (timgraham) [inline django/db/models/query.py]: Check that

File: django/db/models/query.py [hunk 9]
```diff
@@ -423,13 +423,18 @@ def bulk_create(self, objs, batch_size=None):
...
+        # model detects the inheritance pattern ConcreteGrandParent ->
```
Comment (timgraham) [inline django/db/models/query.py]: . Detects

File: tests/bulk_create/tests.py [hunk 10]
```diff
@@ -35,22 +38,48 @@ def test_efficiency(self):
...
+            ValueError, "Can't bulk create a multi-table inherited model",
+            *args, **kwargs)
+
+    def test_multi_table_inheritance_unsupported(self):
         Restaurant.objects.bulk_create([
```
Comment (timgraham) [inline tests/bulk_create/tests.py]: I'm confused why queries that do work are in the "unsupported" test methods?

**Comment by wkschwartz** [comment]
Done responding to the last set of comments. Let me know what else needs fixing.

File: tests/bulk_create/tests.py [hunk 11]
```diff
@@ -35,21 +38,39 @@ def test_efficiency(self):
...
-        ], attrgetter("name"))
+        with self.assertRaisesMessage(ValueError, expected_message):
+            ProxyMultiCountry.objects.bulk_create([
+                ProxyMultiCountry(name="Fillory", iso_two_letter="FL")
+            ])
+        with self.assertRaisesMessage(ValueError, expected_message):
+            ProxyMultiProxyCountry.objects.bulk_create([
+                ProxyMultiProxyCountry(name="Fillory", iso_two_letter="FL")
+            ])
+        expected_empty_models = (Pizzeria, Restaurant, Country, ProxyCountry,
```
Comment (timgraham) [inline tests/bulk_create/tests.py]: is this to ensure that `bulk_create()` didn't create any objections even though it raised an error? I don't think that's really needed. Otherwise, this patch looks good.

Comment (wkschwartz) [inline tests/bulk_create/tests.py]: Yes. This test prevents a regression in which the ValueError gets moved in the code to after where the QuerySet tries to execute the write and can only get part way through it.

I think having this here is important, but ultimately it's your code base. Let me know if removing it is necessary to getting this merged.

**Comment by wkschwartz** [comment]
If I'm not removing that assertion, do you want me to squash my commits now?

File: tests/bulk_create/tests.py (hunk 11 above)
Comment (timgraham) [inline tests/bulk_create/tests.py]: I don't think such a regression is likely and we don't have similar checking elsewhere so I think it's fine to remove.
Yes, please squash commits.

**Comment by wkschwartz** [comment]
All squashed. Let me know if there's anything else that needs to happen for this to get merged.

**Comment by timgraham** [comment]
merged in 9a5cfa05a0d889db9daa7872a9697daa883f3609, thanks!
